src/lib.rs

```diff
@@ -1,202 +1,284 @@
+
+use anyhow::{anyhow, Result};
+use boringtun::crypto::{X25519PublicKey, X25519SecretKey};
+use pyo3::exceptions::{PyKeyError, PyOSError, PyValueError};
 use pyo3::prelude::*;
+use pyo3::types::{PyBytes, PyString, PyTuple};
 use tokio::sync::mpsc;
+use tokio::sync::mpsc::error::SendError;
+use tokio::sync::mpsc::{channel, unbounded_channel};
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::error::RecvError;
 use tokio::task::JoinHandle;
+use crate::messages::{ConnectionId, TransportCommand, TransportEvent};
 
-type ConnectionId = u32;
-
-#[derive(Clone, Debug)]
-struct PySockAddr(SocketAddr);
-
-impl Display for PySockAddr {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        self.0.fmt(f)
-    }
-}
-
-impl IntoPy<PyObject> for PySockAddr {
-    fn into_py(self, py: Python<'_>) -> PyObject {
-        match self.0 {
-            SocketAddr::V4(addr) => (addr.ip().to_string(), addr.port()).into_py(py),
-            SocketAddr::V6(addr) => {
-                log::debug!(
-                    "converting ipv6 to python, not sure if this is correct: {:?}",
-                    (addr.ip().to_string(), addr.port())
-                );
-                (addr.ip().to_string(), addr.port()).into_py(py)
-            },
-        }
-    }
-}
+mod tcp;
+mod wireguard;
+mod messages;
+mod virtual_device;
 
 #[pyclass]
-#[derive(Debug)]
-struct ConnectionEstablished {
-    #[pyo3(get)]
+struct TcpStream {
     connection_id: ConnectionId,
-    #[pyo3(get)]
-    src_addr: PySockAddr,
-    #[pyo3(get)]
-    dst_addr: PySockAddr,
+    event_tx: mpsc::UnboundedSender<TransportCommand>,
+    peername: SocketAddr,
+    sockname: SocketAddr,
+    original_dst: SocketAddr,
 }
 
 #[pymethods]
-impl ConnectionEstablished {
-    fn __repr__(&self) -> String {
-        format!(
-            "ConnectionEstablished({}, {}, {})",
-            self.connection_id, self.src_addr, self.dst_addr
-        )
+impl TcpStream {
+    fn read<'p>(&self, py: Python<'p>, n: u32) -> PyResult<&'p PyAny> {
+        let (tx, rx) = oneshot::channel();
+        self.event_tx
+            .send(TransportCommand::ReadData(self.connection_id, n, tx))
+            .map_err(event_queue_unavailable)?;
+        pyo3_asyncio::tokio::future_into_py(py, async move {
+            let data = rx.await.map_err(connection_closed)?;
+            let bytes: Py<PyBytes> = Python::with_gil(|py| PyBytes::new(py, &data).into_py(py));
+            Ok(bytes)
+        })
     }
-}
-
-
-#[pyclass]
-#[derive(Debug)]
-struct DataReceived {
-    #[pyo3(get)]
-    connection_id: ConnectionId,
-    #[pyo3(get)]
-    data: Vec<u8>,
-}
 
-#[pymethods]
-impl DataReceived {
-    fn __repr__(&self) -> String {
-        format!("DataReceived({}, {:x?})", self.connection_id, self.data)
+    fn write(&self, data: Vec<u8>) -> PyResult<()> {
+        self.event_tx
+            .send(TransportCommand::WriteData(self.connection_id, data))
+            .map_err(event_queue_unavailable)?;
+        Ok(())
```

Comment:
This is an annoying quirk of the existing Python Stream API (which we cannot change): `writer.write()` has an unbounded buffer and never blocks. So we need this to be unbounded in Rust-land as well.